### PR TITLE
Fix introduce var duplicating newlines

### LINKED ISSUE #713: Extra empty lines when introducing new variable
Before:
```
fn foo() -> bool {

    
    return <|>true<|>;
}
```
After:
```
fn foo() -> bool {

    
    let <|>var_name = true;

    
    return var_name;
}
```

That's because of this:
https://github.com/rust-analyzer/rust-analyzer/blob/28fdb8d03caf1ab8b40ed0fcbe8e47451fe030d9/crates/ra_ide_api_light/src/assists/introduce_variable.rs#L36

An easy fix would be considering only the last spaces of that WHITESPACE block.

This should be an easy issue for someone to pick-up

### PR DESCRIPTION
**Pull request by vipentti**
This fixes #713.

If the block before the statement we want to use introduce var on, had empty
lines these empty lines would also be added between the let-statement and
the current line where the new variable is used.

This fixes that by trimming excess newlines from the start of the indent chunk
and simply adding a single newline (when the chunk had newlines) between the
let-statement and the current statement. If there were no newlines this
matches the previous behaviour.

### CODE REVIEW

File: crates/ra_assists/src/introduce_variable.rs [hunk 1]
```diff
@@ -44,7 +44,22 @@ pub(crate) fn introduce_variable(ctx: AssistCtx<impl HirDatabase>) -> Option<Ass
...
+            // but we do not want to duplicate possible
+            // extra newlines in the indent block
+            for chunk in indent.text().chunks() {
+                if chunk.starts_with("\r\n") {
+                    buf.push_str("\r\n");
+                    buf.push_str(chunk.trim_start_matches("\r\n"));
+                } else if chunk.starts_with("\n") {
+                    buf.push_str("\n");
+                    buf.push_str(chunk.trim_start_matches("\n"));
+                } else {
```
Comment (kjeremy) [inline crates/ra_assists/src/introduce_variable.rs]: I've seen this in a few spots where we look for `\r\n` or `\n`. Is there a crate (something in std?) that can help with terminators?

**Comment by kjeremy** [comment]
Other than my one question LGTM

File: crates/ra_assists/src/introduce_variable.rs (hunk 1 above)
Comment (vipentti) [reply crates/ra_assists/src/introduce_variable.rs]: I tried to look for something in the std or any crates that would help dealing with terminators but could not find any.

Comment (matklad) [reply crates/ra_assists/src/introduce_variable.rs]: I think we should ideally convert \r\n to \n somewhere on the VFS layer, or maybe when we are constructing a syntax tree. Internals should always use \n as a line separator.

Will make an issue for that.

**Comment by matklad** [comment]
bors r+

Thanks!

File: crates/ra_assists/src/introduce_variable.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_assists/src/introduce_variable.rs]: https://github.com/rust-analyzer/rust-analyzer/issues/775

**Pull request merged**